### ISSUE TITLE
join without c_str could produce unforeseen null character in string

### ISSUE THREAD

**Issue by saki7** [ericniebler/range-v3#1403]
code:

```cpp
std::vector<std::string> strings{"aaa", "bbb", "ccc"};
auto const str_a = strings | ranges::views::join("__") | ranges::to<std::string>(); // (1)
auto const str_b = strings | ranges::views::join(ranges::views::c_str("__")) | ranges::to<std::string>(); // (2)
std::cout << "str_a: " << str_a << ", size: " << str_a.size() << std::endl;
std::cout << "str_b: " << str_b << ", size: " << str_b.size() << std::endl;
```

prints:
```
str_a: aaa__bbb__ccc, size: 15
str_b: aaa__bbb__ccc, size: 13
```

The former line (1) embeds null (`\0`) character after every delimiters because of the lack of `ranges::views::c_str(...)`. 

While I *could* assume that this output is pretty much the expected behavior since __the string is actually a range of characters__, when used in the context of ranges. However this could become a pitfall when one's using this library to format strings (and I believe that it is quite common use-case).

The mistaken result is easy to produce and hard to debug since null characters are never printed to terminal. To solve this, we could either:
- (1) Mark some overloads of  `views::join()` `delete`d for disambiguation, or
- (2) Show some kind of warnings to prevent faulty string concatenation.

**Comment by saki7** [comment]
FYI: `boost::algorithm::join()` produces expected result even when string literals (or c strings) are passed as the delimiter <https://www.boost.org/doc/libs/1_72_0/doc/html/boost/algorithm/join.html>

**Comment by dmsteck** [comment]
I am confused by this though. It is my understanding that a `std::string`, interpreted as a half-open `[begin, end)` character range, should *not* include the null character. The null character is `*str.end()`.

https://ideone.com/tVGKt3

**Comment by johelegp** [comment]
Yes. But there's also another `'\0'`, which is the third element in `"__"`. `views::c_str("__")` only has two elements. `std::string` supports `'\0'` anywhere in its characters.
